server: properly authenticate some admin RPC requests

knz opened the pull request:
Needed for #42520.
Fixes the vulnerability described in #42567 (but does not durably prevent it from creeping back, so more work is needed to fully close the issue).

The the individual commits for details.

knz committed:
server: perform authorization for SQL ran by admin RPCs

Prior to this patch, all the admin RPCs would perform
whatever SQL operations they needed to do using the root user.
See issue https://github.com/cockroachdb/cockroach/issues/42567.

Moreover, because of this the "UI data" payload containing all the UI
configuration settings was shared between all users.

This was possibly a security flaw, as it could enable any
authenticated user (eg via a certificate on the CLI, or via the web
UI) to access SQL data otherwise only available to root.

This patch fixes **part of** issue #42567 by
authorizing the SQL execution of admin RPC endpoints properly.

It also introduces a new SQL built-in function
`crdb_internal.is_admin()` used internally by the admin user.

**The following problem from #42567 is not fixed: RPCs that do not use
SQL are still allowed to proceed as root regardless of the
authenticated user.** This will be addressed by a subsequent commit.

Release note (admin ui change): Certain web UI pages (like the list of
databases, tables) now restrict their content to match the privileges
of the logged-in user.

Release note (admin ui change): The event log now presents all cluster
settings changes, unredacted, when an admin user uses the page.

Release note (admin ui change): Customization of the UI by
users is only properly saved if the user has write privilege
to `system.ui` (i.e. admin users). Also, all authenticated
users share the same customizations. This is a known limitation
and should be lifted in a future version.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/42563)
<!-- Reviewable:end -->

knz commented:
@rolandcrosby @bdarnell do you know if the current/previous behavior was intentional? If not, perhaps this may warrant a security announcement?

rolandcrosby commented:
@piyush-singh can you speak to the intended admin UI behavior and whether this was by design or not?

knz commented:
Filed a separate issue  #42567 to describe the problem.

mberhault commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -3284,6 +3284,31 @@ may increase either contention or retry errors, or both.`,
 
+	// Returns true iff the current user has admin role.
+	// Note: it would be a privacy leak to extend this to check arbitrary usernames.
+	"crdb_internal.is_admin": makeBuiltin(
```
Nice builtin!

mberhault reviewed (commented):
Did anything stop working on the admin UI? List of databases/tables/users? Some of those are unlikely to work as non-admin users.

knz reviewed (commented):
> Did anything stop working on the admin UI? List of databases/tables/users? Some of those are unlikely to work as non-admin users.

I checked manually: the databases/tables/users list now only display what a user would be able to see in SQL otherwise (intended feature). The statements page is empty (this is a admin-only feature, so intended behavior).

The jobs page reports an error, but again I think that's OK since listing all jobs is also admin-only.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42563#-:-Lu3c8Zk2FdYDISm-ZOR:b-8yklm3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @knz, and @mberhault)

___
*[pkg/sql/logictest/testdata/logic_test/crdb_internal, line 616 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42563#-Lu2p6eB5CDlEDDGQguU-r1-616:-Lu3cQHV1ympb0LKDwtR:b-qfw0iz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e1d3ebfcfb4540aa8935de3d50f32b73e3540f3/pkg/sql/logictest/testdata/logic_test/crdb_internal#L616)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

Can you grant `admin` role membership to `testuser` and try again? It should now be true. Or would that need to be done in the CCL logictests?
</blockquote></details>

Done (in ccllogictest).


<!-- Sent from Reviewable.io -->

knz commented:
Discussed offline: this PR is not a complete fix for #42567 but it's moving the needle forward and also it's a prerequisite to #42520 which I'd also like to get off my plate.

So I'd like to motion to merge this even though it's not a complete fix yet.

piyush-singh commented:
This was mostly just unimplemented on our end. We intended to add Authz checks after the addition of Authn in the 2.1 release. Requiring Authn was already a strict improvement as previously the HTTP endpoints were completely unsecured for the admin UI in secure clusters. However before we were able to actually make progress on adding Authz checks, the team was dissolved. Thank you @knz for picking this back up, I should probably have discussed this as a security issue more publicly after the team was dissolved.

mberhault reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42563)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @knz)



<!-- Sent from Reviewable.io -->

bdarnell reviewed (approved):
> this PR is not a complete fix for #42567 

What's left?

> Reviewed 18 of 18 files at r1, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42563#-:-LuF0Xtt6DtKOi54CdcZ:b-dcm847)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)



<!-- Sent from Reviewable.io -->

knz commented:
> What's left?

The code currently works by extracting the authenticated username and passing it to the SQL queries, and letting SQL validate privileges.

There are two problems with this:

- some SQL-based endpoints _work_ with a non-root username but we may not _want_ them to, because they are cluster-wide expensive (risk of DoS) or they contain a non-SQL part which can reveal sensitive information. An example of this is `DataDistribution`.

- Not all endpoints use SQL in the first place. All those that don't are currently still unauthenticated and function as `root`. 

The following endpoints remain problematic in either of these two ways:

- `Liveness`, data leak, it can reveal IP addresses of nodes
- `Drain` - DoS
- `DecommissionStatus` - data leak
- `Decommission` - DoS
- `DataDistribution` - DoS and data leak
- `EnqueueRange` - data leak: you give it any range ID and it gives you back a detailed trace including key strings full of PII
- any additional admin RPC added in the future that does not use SQL, this PR does not add any guardrail to ensure the future engineer does the right thing.

Also see my additional comment on the issue.

knz committed:
serverpb: don't expose Drain() via HTTP

The Drain() (`cockroach quit`) RPC is used over gRPC by `cockroach
quit`. There is no requirement to have it exposed via HTTP, and
moreover it is a security problem (DoS) waiting to happen
since the request is not fully authorized.

Release note: none

knz committed:
server: authorize more requests

Previously, admin/status RPCs available over HTTP but that did not
perform SQL operations would not verify authorization from
clients (the only authorization was that performed by SQL privilege
checks).

This patch fixes it.

The following services are now blocked from non-admin users:

- `TableStats`. This is a temporary limitation, until this RPC learns
  how to verify privileges on the specific table for which stats are
  requested.
- `NonTableStats`
- `DataDistribution`
- `EnqueueRange`
- `RangeLog`
- `Gossip`
- `EngineStats`
- `Allocator`
- `AllocatorRange`
- `Certificates`
- `Details`
- `GetFiles`
- `LogFilesList`
- `LogFile`
- `Logs`
- `Stacks`
- `Profile`
- `RaftDebug`
- `Ranges`
- `HotRanges`
- `Range`
- `SpanStats`
- `Stores`

For reference, the following services were already properly authorized
by SQL privilege checks:

- `Databases`
- `DatabaseDetails`
- `TableDetails`
- `Users`
- `Events`
- `Jobs`
- `Locations`
- `QueryPlan`

The following were already authorized using a custom check:

- `ListSessions` (user must be either admin or can only list their own
  sessions)
- `ListLocalSessions` (ditto)

The following are now authorized using a custom check:

- `CancelSession` (user must be either admin or operate on their own sessions)
- `CancelQuery` (user must be either admin or operate on their own queries)

The following RPCs are not authorized but do not leak sensitive
information:

- `Health`
- `Liveness`
- `Settings`

Release note (admin ui change): Access to table statistics are
temporary blocked from access by non-admin users until further notice,
for security reasons.

Release note (admin ui change): Certain debug pages have been blocked
from non-admin users for security reasons.

knz commented:
I have investigated all the remaining RPCs, closed the remaining gaps, documented the UX changes in release notes and left TODOs for the remaining work. So this means the PR will remove the "vulnerability" part of #42567.

I still think the issue should remain open until we have a better mechanism (common check with whitelist) and more tests! But at least we can get this through for the release.

knz commented:
RFAL, but I won't look further until Monday.

bdarnell reviewed (approved):
> Reviewed 5 of 5 files at r4, 3 of 3 files at r5, 2 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42563)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)



<!-- Sent from Reviewable.io -->

knz commented:
Thanks!

bors r+

knz commented:
Transient upstream pip error in ORM tests

bors r+

knz commented:
@bdarnell it's quite a hassle to do port this to 2.1, essentially a lot of the pieces have to be re-done from scratch due to arch changes in the 2.1->19.1 transition. Should I still look into it?
